lib/rex/proto/dns/resolver.rb

```diff
@@ -375,8 +369,95 @@ def query(name, type = Dnsruby::Types::A, cls = Dnsruby::Classes::IN)
       return send(name,type,cls)
 
     end
+  end # Resolver
 
+  class CachedResolver < Resolver
+    include Rex::Proto::DNS::Constants
+    attr_accessor :cache
+
+    #
+    # Initialize resolve with cache
+    #
+    # @param config, Hash, Resolver config
+    def initialize(config = {})
+      super(config)
+      self.cache = Rex::Proto::DNS::Cache.new
+      # Read hostsfile into cache
+      hf = Rex::Compat.is_windows ? '%WINDIR%/system32/etc/hosts' : '/etc/hosts'
+      entries = File.read(hf).lines.map(&:strip).select do |entry|
+        Rex::Socket.is_ip_addr?(entry.gsub(/\s+/,' ').split(' ').first) and
+        not entry.match(/::.*ip6-/)
```

Comment:
Alright on second review and discussions internally I'd much rather we do `Rex::Socket.is_ipv6?` than try do regex like this. It will make maintaining this a hell of a lot easier and prevent potential issues. Assume you were trying to identify IPv6 addresses here.

Reply:
I can make the conversion, but this is copied from older code and is _much_ more efficient than a regex match:
```ruby
3.1.2 :005 > Benchmark.realtime { (1..10000).to_a.each { address.include?(':') } }
 => 0.0006861269939690828 
3.1.2 :006 > Benchmark.realtime { (1..10000).to_a.each { address.match?(MATCH_IPV6) } }
 => 0.01085773401428014 
```

Reply:
Yep that was on Ubuntu, I'll see if I can try take a look at this again.

Reply:
Okay you'll have to forgive me for asking this but I'm still confused why we are explicitly forbidding IPv6 addresses here?

Reply:
Not at all, this is some deep-level stuff and the more of it we doc here, the less chance of lost lore later.
So, that line doesn't ignore IPv6 entries, it ignores the ubuntu notations for IPv6 addr spaces:
```
# The following lines are desirable for IPv6 capable hosts
::1 ip6-localhost ip6-loopback
fe00::0 ip6-localnet
ff00::0 ip6-mcastprefix
ff02::1 ip6-allnodes
ff02::2 ip6-allrouters
ff02::3 ip6-allhosts
```

Reply:
@sempervictus So is the goal to avoid the new IPv6 addresses that are indicative of ranges such as localnet etc since IPv6 added some addresses for that sort of thing? Just concerned that this will skip over stuff like if we have multiple hostnames for one address such as if `::1` was also set to `localhost` as a name. Additionally concerned that the current behavior of IPv4 is to keep such addresses such as `localhost` and this might end up becoming a discrepancy between IPv4 and IPv6.

Reply:
Sounds good, sorry for my confusion here I must have been sleep deprived or just not reading the code right as this makes sense now.

If we can just add in the comment for documentations sake I'll be happy to close this as resolved.

Reply:
Done

Reply:
Comment verified, marking this as resolved.